services/hh/service.go

```diff
@@ -56,55 +61,81 @@ func NewService(c Config, w shardWriter, m metaStore) *Service {
 	key := strings.Join([]string{"hh", c.Dir}, ":")
 	tags := map[string]string{"path": c.Dir}
 
-	s := &Service{
-		cfg:     c,
-		statMap: influxdb.NewStatistics(key, "hh", tags),
-		Logger:  log.New(os.Stderr, "[handoff] ", log.LstdFlags),
-	}
-	processor, err := NewProcessor(c.Dir, w, m, ProcessorOptions{
-		MaxSize:        c.MaxSize,
-		RetryRateLimit: c.RetryRateLimit,
-	})
-	if err != nil {
-		s.Logger.Fatalf("Failed to start hinted handoff processor: %v", err)
+	return &Service{
+		cfg:         c,
+		closing:     make(chan struct{}),
+		processors:  make(map[uint64]*NodeProcessor),
+		statMap:     influxdb.NewStatistics(key, "hh", tags),
+		Logger:      log.New(os.Stderr, "[handoff] ", log.LstdFlags),
+		shardWriter: w,
+		metastore:   m,
 	}
-
-	processor.Logger = s.Logger
-	s.HintedHandoff = processor
-	return s
 }
 
 func (s *Service) Open() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	if !s.cfg.Enabled {
-		// Allow Open to proceed, but don't anything.
+		// Allow Open to proceed, but don't do anything.
 		return nil
 	}
-
 	s.Logger.Printf("Starting hinted handoff service")
-
-	s.mu.Lock()
-	defer s.mu.Unlock()
-
 	s.closing = make(chan struct{})
 
+	// Register diagnostics if a Monitor service is available.
+	if s.Monitor != nil {
+		s.Monitor.RegisterDiagnosticsClient("hh", s)
+	}
+
+	// Create the root directory if it doesn't already exist.
 	s.Logger.Printf("Using data dir: %v", s.cfg.Dir)
+	if err := os.MkdirAll(s.cfg.Dir, 0700); err != nil {
+		return fmt.Errorf("mkdir all: %s", err)
+	}
+
+	// Create a node processor for each node directory.
+	files, err := ioutil.ReadDir(s.cfg.Dir)
+	if err != nil {
+		return err
+	}
 
-	s.wg.Add(3)
-	go s.retryWrites()
-	go s.expireWrites()
-	go s.deleteInactiveQueues()
+	for _, file := range files {
+		nodeID, err := strconv.ParseUint(file.Name(), 10, 64)
+		if err != nil {
+			// Not a number? Skip it.
+			continue
+		}
+
+		n := NewNodeProcessor(nodeID, s.pathforNode(nodeID), s.shardWriter, s.metastore)
+		if err := n.Open(); err != nil {
+			return err
+		}
+		s.processors[nodeID] = n
+	}
+
+	s.wg.Add(1)
+	go s.purgeInactiveProcessors()
 
 	return nil
 }
 
 func (s *Service) Close() error {
+	s.Logger.Println("shutting down hh service")
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	for _, p := range s.processors {
+		if err := p.Close(); err != nil {
+			return err
+		}
+	}
+
 	if s.closing != nil {
 		close(s.closing)
 	}
 	s.wg.Wait()
+	s.closing = nil
```

Comment:
Setting this to `nil` prevents existing goroutines from selecting it.  That has created issues in other parts of the code.